### PR TITLE
Add SNI support to the apiserver

### PR DESCRIPTION
**Pull request by sttts**
This PR adds the `--tls-sni-key-cert` flag to the apiserver. It can be passed multiple times in the following ways:

``` shell
$ apiserver \
    --tls-sni-cert-key '*.example.com,example.com: example.key,example.crt' \
    --tls-sni-cert-key 'foo.key,foo.crt'
```

The first variant explicitly sets the accepted domain names, the second variant reads the common names and DNS names from the certificate itself.

If no domain name matches, the existing certificate (`--tls-cert-file`) is used.

``` golang
    fs.Var(config.NewNamedCertKeyArray(&s.SNICertKeys), "tls-sni-cert-key", ""+
        "A pair of x509 certificate and private key file paths, optionally prefixed with a list of "+
        "domain patterns which are fully qualified domain names, possibly with prefixed wildcard "+
        "segments. If no domain patterns are provided, the names of the certificate are "+
        "extracted. Non-wildcard matches trump over wildcard matches, explicit domain patterns "+
        "trump over extracted names. For multiple key/certificate pairs, use the "+
        "--tls-sni-key-cert multiple times. "+
        "Examples: \"example.key,example.crt\" or \"*.foo.com,foo.com:foo.key,foo.crt\".")
```

``` release-note
Add SNI support to the apiserver

Pass multiple certificates and domain name patterns with `--tls-sni-cert-key` and the right certificate will be chosen depending on the url the client is using.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35109)

### CODE REVIEW

File: pkg/genericapiserver/genericapiserver.go [hunk 1]
```diff
@@ -547,3 +553,40 @@ func waitForSuccessfulDial(https bool, network, address string, timeout, interva
...
+		return nil, nil
+	}
+	namedCerts := map[string]*tls.Certificate{}
+	for _, nkc := range namedKeyCerts {
+		cert, err := tls.LoadX509KeyPair(nkc.CertFile, nkc.KeyFile)
+		if err != nil {
+			return nil, err
+		}
+		if len(nkc.Names) > 0 {
+			for _, name := range nkc.Names {
```
Comment (liggitt) [inline pkg/genericapiserver/genericapiserver.go]: explicitly specified names should trump auto-populated names

for example, if I have the following certs:
cert1: CN=bar.com
cert2: CN=foo.com, dnsSAN=bar.com

and I specify these flags:
`--tls-sni-cert-key=cert1.pem,key1.pem,bar.com --tls-sni-cert-key=cert2.pem,key2.pem`

I should end up with
"bar.com": cert1
"foo.com": cert2

File: pkg/genericapiserver/options/server_run_options.go [hunk 2]
```diff
@@ -473,13 +474,20 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
...
 	fs.StringVar(&s.TLSPrivateKeyFile, "tls-private-key-file", s.TLSPrivateKeyFile,
-		"File containing x509 private key matching --tls-cert-file.")
+		"File containing the default x509 private key matching --tls-cert-file.")
+
+	fs.Var(config.NewNamedKeyCertArray(&s.SNIKeyCerts), "tls-sni-key-cert", ""+
+		"A pair of x509 key and certificate file paths, optionally prefixed with a list of "+
+		"domain patterns which are fully qualified domain names, possibly with prefixed wildcard "+
+		"segments. If no domain patterns are provided, the common names of the certificate are "+
+		"used instead. For multiple key/certificate pairs, use the --tls-sni-key-cert multiple times. "+
+		"Examples: \"example.key,example.crt\" or \"*.foo.com,foo.com:foo.key,foo.crt\".")
```
Comment (liggitt) [inline pkg/genericapiserver/options/server_run_options.go]: maybe put the variable args at the end, and the cert first (all the go stdlib functions do cert,key)? (e.g. `cert,key[,name,...]`)

File: pkg/util/config/namedkeycert_flag.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		*a.value = append(*a.value, nkc)
+	}
+	return nil
+}
+
+func (a *NamedKeyCertArray) Type() string {
+	return "namedKeyCert"
+}
+
+func (a *NamedKeyCertArray) String() string {
```
Comment (liggitt) [inline pkg/util/config/namedkeycert_flag.go]: this is for display only, not for re-parsing, right?

File: pkg/genericapiserver/genericapiserver.go [hunk 4]
```diff
@@ -200,7 +206,8 @@ func (s *GenericAPIServer) Run() {
 				// Can't use SSLv3 because of POODLE and BEAST
 				// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
 				// Can't use TLSv1.1 because of RC4 cipher usage
-				MinVersion: tls.VersionTLS12,
+				MinVersion:        tls.VersionTLS12,
+				NameToCertificate: namedCerts,
```
Comment (liggitt) [inline pkg/genericapiserver/genericapiserver.go]: does this need to include the names from the default cert as well?

Comment (liggitt) [inline pkg/genericapiserver/genericapiserver.go]: when I did this in OpenShift, I had to use a custom `GetCertificate` function instead of `NameToCertificate`:

> // Set SNI certificate func
> // Do not use NameToCertificate, since that requires certificates be included in the server's tlsConfig.Certificates list,
> // which we do not control when running with http.Server#ListenAndServeTLS
> GetCertificate: cmdutil.GetCertificateFunc(extraCerts),

can you verify this is no longer the case?

edit: looks like this may have been fixed in go1.5

**Comment by sttts** [comment]
@k8s-bot gci gke e2e test this

**Comment by liggitt** [comment]
needs tests that actually start a TLS server and check that a default-cert+sni-certs config actually serves the sni cert to one of the sni hostnames, and falls back to the default cert for all other hostnames

**Review by liggitt — CHANGES_REQUESTED** [review]

**Comment by sttts** [comment]
@liggitt do we have any test for `GenericApiServer.run()` right now? I guess we don't, but we should. I vote for adding a close channel to `run()` and allow clean shutdown. /cc @deads2k

File: pkg/genericapiserver/genericapiserver.go (hunk 4 above)
Comment (liggitt) [inline pkg/genericapiserver/genericapiserver.go]: there was a default GetCertificate in go1.4 as well, which behaved differently, and required jumping through hoops to get the right cert selected. The current implementation looks like it does what we want, but I do want tests to make sure the combination of ListenAndServeTLS(defaultCert,defaultKey) and the NameToCertificate map we're passing in selects the right cert.

**Comment by liggitt** [comment]
I don't see a clean way to shut down a server started with ListenAndServeTLS

File: pkg/genericapiserver/options/server_run_options.go (hunk 2 above)
Comment (liggitt) [inline pkg/genericapiserver/options/server_run_options.go]: sure, `cert:key[:name,...]`?

Comment (sttts) [inline pkg/genericapiserver/options/server_run_options.go]: key,cert:name,name,name

Comment (sttts) [inline pkg/genericapiserver/options/server_run_options.go]: I am fine with either order. Where do we have cert,key? I my head the key exists first, then the cert ;)

Comment (liggitt) [inline pkg/genericapiserver/options/server_run_options.go]: server.ListenAndServeTLS(cert,key)
tls.X509KeyPair(certbytes,keybytes)
tls.LoadX509KeyPair(cert,key)

Comment (sttts) [inline pkg/genericapiserver/options/server_run_options.go]: That's not user facing. But cert/key is fine as well.

File: pkg/genericapiserver/genericapiserver.go (hunk 1 above)
Comment (sttts) [inline pkg/genericapiserver/genericapiserver.go]: I think we should give auto-generation precedence, but enforce that the first cert given trumps any later one. That matches your example. WDYT?

**Comment by deads2k** [comment]
> @liggitt do we have any test for GenericApiServer.run() right now? I guess we don't, but we should. I vote for adding a close channel to run() and allow clean shutdown. /cc @deads2k

I'm ok adding a channel, but I doubt it will bring everything down cleanly.  We had a lot of trouble with that in the distant past.

**Comment by sttts** [comment]
@deads2k former go versions (<= 1.3) had races when closing the `http.Listener` asynchronously to bring down a server. I have read this has improved and maybe this was a reason for the trouble your are mentioning.

File: pkg/genericapiserver/genericapiserver.go (hunk 1 above)
Comment (liggitt) [inline pkg/genericapiserver/genericapiserver.go]: explicitly given names should always outweigh auto-detected ones

Comment (sttts) [inline pkg/genericapiserver/genericapiserver.go]: done

**Comment by sttts** [comment]
@deads2k @liggitt added integration tests

File: pkg/genericapiserver/options/server_run_options.go [hunk 5]
```diff
@@ -473,13 +474,22 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
...
-		"File containing x509 private key matching --tls-cert-file.")
+		"File containing the default x509 private key matching --tls-cert-file.")
+
+	fs.Var(config.NewNamedCertKeyArray(&s.SNICertKeys), "tls-sni-cert-key", ""+
+		"A pair of x509 certificate and private key file paths, optionally prefixed with a list of "+
+		"domain patterns which are fully qualified domain names, possibly with prefixed wildcard "+
+		"segments. If no domain patterns are provided, the names of the certificate are "+
+		"extracted. Non-wildcard matches trump over wildcard matches, explicit domain patterns "+
+		"trump over extracted names. For multiple key/certificate pairs, use the "+
+		"--tls-sni-key-cert multiple times. "+
```
Comment (ncdc) [inline pkg/genericapiserver/options/server_run_options.go]: You mean `--tls-sni-cert-key`, right?

File: pkg/genericapiserver/genericapiserver.go [hunk 6]
```diff
@@ -272,7 +226,9 @@ func (s preparedGenericAPIServer) Run() {
 		glog.Errorf("Unable to send systemd daemon successful start message: %v\n", err)
 	}
 
-	select {}
+	select {
```
Comment (ncdc) [inline pkg/genericapiserver/genericapiserver.go]: Instead of the select block, can't you just do `<-stopCh`?

File: pkg/genericapiserver/serve.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+func runServer(server *http.Server, stopCh <-chan struct{}) (int, error) {
+	addr := server.Addr
+	if addr == "" {
+		addr = ":https"
+	}
+
+	// first listen is synchronous (fail early!)
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return 0, fmt.Errorf("Failed to listen on %v: %v", addr, err)
```
Comment (ncdc) [inline pkg/genericapiserver/serve.go]: lowercase failed

File: pkg/genericapiserver/serve.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+	*net.TCPListener
+}
+
+func (ln tcpKeepAliveListener) Accept() (c net.Conn, err error) {
+	tc, err := ln.AcceptTCP()
+	if err != nil {
+		return
+	}
+	tc.SetKeepAlive(true)
+	tc.SetKeepAlivePeriod(3 * time.Minute)
```
Comment (ncdc) [inline pkg/genericapiserver/serve.go]: Is it ok to hard-code this value? Maybe at least make it a constant if it's not configurable?

File: pkg/genericapiserver/options/server_run_options.go (hunk 5 above)
Comment (sttts) [inline pkg/genericapiserver/options/server_run_options.go]: Right, this name changed several times. This is the previous name.

File: pkg/genericapiserver/serve.go (hunk 8 above)
Comment (sttts) [inline pkg/genericapiserver/serve.go]: Or even a parameter. This is only moved code though.

File: pkg/util/config/namedcertkey_flag.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+// NamedCertKeyArray is a flag value parsing NamedKeyCerts, each passed with its own
+// flag instance (in contrast to comma separated slices).
+type NamedCertKeyArray struct {
+	value   *[]NamedCertKey
+	changed bool
+}
+
+var _ flag.Value = &NamedCertKey{}
+
+// NewNamedKeyCertArray creates a new NamedCertKeyArray with the internal value
```
Comment (sttts) [inline pkg/util/config/namedcertkey_flag.go]: s/NamedKeyCerts/NamedCertKeys/

File: pkg/genericapiserver/options/server_run_options.go [hunk 10]
```diff
@@ -473,13 +474,22 @@ func (s *ServerRunOptions) AddUniversalFlags(fs *pflag.FlagSet) {
...
 		"after server cert). If HTTPS serving is enabled, and --tls-cert-file and "+
 		"--tls-private-key-file are not provided, a self-signed certificate and key "+
 		"are generated for the public address and saved to /var/run/kubernetes.")
 
 	fs.StringVar(&s.TLSPrivateKeyFile, "tls-private-key-file", s.TLSPrivateKeyFile,
-		"File containing x509 private key matching --tls-cert-file.")
+		"File containing the default x509 private key matching --tls-cert-file.")
+
+	fs.Var(config.NewNamedCertKeyArray(&s.SNICertKeys), "tls-sni-cert-key", ""+
+		"A pair of x509 certificate and private key file paths, optionally prefixed with a list of "+
```
Comment (liggitt) [inline pkg/genericapiserver/options/server_run_options.go]: s/prefixed/suffixed/

File: pkg/genericapiserver/genericapiserver.go [hunk 11]
```diff
@@ -192,70 +192,19 @@ func (s *GenericAPIServer) PrepareRun() preparedGenericAPIServer {
...
-			for {
-				if err := insecureServer.ListenAndServe(); err != nil {
-					glog.Errorf("Unable to listen for insecure (%v); will try again.", err)
-				}
-				time.Sleep(15 * time.Second)
-			}
-		}()
 	}
 
 	// Attempt to verify the server came up for 20 seconds (100 tries * 100ms, 100ms timeout per try) per port
```
Comment (liggitt) [inline pkg/genericapiserver/genericapiserver.go]: if the initial listen is synchronous now, I think the waitForSuccessfulDial calls can be removed

File: pkg/genericapiserver/serve.go [hunk 12]
```diff
@@ -20,8 +20,162 @@ import (
...
+		Handler:        s.Handler,
+		MaxHeaderBytes: 1 << 20,
+		TLSConfig: &tls.Config{
+			// Can't use SSLv3 because of POODLE and BEAST
+			// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+			// Can't use TLSv1.1 because of RC4 cipher usage
+			MinVersion:        tls.VersionTLS12,
+			NameToCertificate: namedCerts,
+			// enable HTTP2 for go's 1.7 HTTP Server
+			NextProtos: []string{"h2", "http/1.1"},
```
Comment (liggitt) [inline pkg/genericapiserver/serve.go]: is there anything else http2-related we miss by setting up our own listener?

File: pkg/genericapiserver/serve.go [hunk 13]
```diff
@@ -20,8 +20,162 @@ import (
...
+			var listener net.Listener
+			listener = tcpKeepAliveListener{ln.(*net.TCPListener)}
+			if server.TLSConfig != nil {
+				listener = tls.NewListener(listener, server.TLSConfig)
+			}
+
+			err := server.Serve(listener)
+			glog.Errorf("Error serving %v (%v); will try again.", addr, err)
+
+			// listen again, but never retry instead of fail.
```
Comment (liggitt) [inline pkg/genericapiserver/serve.go]: "never retry" comment is confusing next to what looks like a retry loop

File: pkg/genericapiserver/serve.go [hunk 14]
```diff
@@ -82,3 +236,21 @@ func getNamedCertificateMap(namedCertKeys []NamedCertKey) (map[string]*tls.Certi
 
 	return tlsCertsByName, nil
 }
+
+// tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
+// connections. It's used by ListenAndServe and ListenAndServeTLS so
+// dead TCP connections (e.g. closing laptop mid-download) eventually
+// go away.
+type tcpKeepAliveListener struct {
```
Comment (liggitt) [inline pkg/genericapiserver/serve.go]: comment where this came from

File: test/integration/discoverysummarizer/discoverysummarizer_test.go [hunk 15]
```diff
@@ -67,6 +67,7 @@ func runDiscoverySummarizer(t *testing.T) string {
 
 func runAPIServer(t *testing.T) string {
 	serverRunOptions := apiserver.NewServerRunOptions()
+	serverRunOptions.SecurePort = 0
```
Comment (liggitt) [inline test/integration/discoverysummarizer/discoverysummarizer_test.go]: can we change this without disabling it?

File: test/integration/federation/server_test.go [hunk 16]
```diff
@@ -86,6 +86,7 @@ var groupVersions = []unversioned.GroupVersion{
 
 func TestRun(t *testing.T) {
 	s := options.NewServerRunOptions()
+	s.SecurePort = 0
```
Comment (liggitt) [inline test/integration/federation/server_test.go]: change without disabling?

File: pkg/util/cert/cert.go [hunk 17]
```diff
@@ -138,7 +138,7 @@ func GenerateSelfSignedCertKey(host string, alternateIPs []net.IP, alternateDNS
 	template := x509.Certificate{
 		SerialNumber: big.NewInt(1),
 		Subject: pkix.Name{
-			CommonName: fmt.Sprintf("%s@%d", host, time.Now().Unix()),
+			CommonName: host,
```
Comment (liggitt) [inline pkg/util/cert/cert.go]: many clients fail TLS connections if they see two different certs with the same serial number from the same signer (and they have a local memory of previously seen certs from a given signer). we're hard-coding serialnumber:1 here, and host is likely to be the same between local cluster setups. that's why the timestamp was added

**Review by liggitt — CHANGES_REQUESTED** [review]

File: pkg/genericapiserver/options/server_run_options.go (hunk 5 above)
Comment (sttts) [inline pkg/genericapiserver/options/server_run_options.go]: done

File: pkg/genericapiserver/genericapiserver.go (hunk 6 above)
Comment (sttts) [inline pkg/genericapiserver/genericapiserver.go]: done

File: pkg/genericapiserver/serve.go (hunk 7 above)
Comment (sttts) [inline pkg/genericapiserver/serve.go]: done

File: pkg/genericapiserver/serve.go (hunk 8 above)
Comment (sttts) [inline pkg/genericapiserver/serve.go]: done

File: pkg/util/config/namedcertkey_flag.go [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+	nkc.CertFile = strings.TrimSpace(cs[0])
+	nkc.KeyFile = strings.TrimSpace(cs[1])
+	return nil
+}
+
+func (*NamedCertKey) Type() string {
+	return "namedCertKey"
+}
+
+// NamedCertKeyArray is a flag value parsing NamedKeyCerts, each passed with its own
```
Comment (sttts) [inline pkg/util/config/namedcertkey_flag.go]: done

File: pkg/util/config/namedcertkey_flag.go (hunk 9 above)
Comment (sttts) [inline pkg/util/config/namedcertkey_flag.go]: done

File: pkg/genericapiserver/serve.go (hunk 12 above)
Comment (sttts) [inline pkg/genericapiserver/serve.go]: Luckily there is golang.org/x/net/http2 which exports everything that is done by `ListenAndServeTLS`.

File: pkg/genericapiserver/serve.go [hunk 19]
```diff
@@ -20,8 +20,169 @@ import (
...
+	s.effectiveInsecurePort, err = runServer(insecureServer, stopCh)
+	return err
+}
+
+// runServer listens on the given port, then spawns a go-routine continuously serving
+// until the stopCh is closed. The port is returned. This function does not block.
+func runServer(server *http.Server, stopCh <-chan struct{}) (int, error) {
+	addr := server.Addr
+	if addr == "" {
+		addr = ":https"
```
Comment (sttts) [inline pkg/genericapiserver/serve.go]: done

File: pkg/genericapiserver/serve.go (hunk 13 above)
Comment (sttts) [inline pkg/genericapiserver/serve.go]: indeed.

File: pkg/genericapiserver/serve.go (hunk 14 above)
Comment (sttts) [inline pkg/genericapiserver/serve.go]: done

File: pkg/genericapiserver/options/server_run_options.go (hunk 10 above)
Comment (sttts) [inline pkg/genericapiserver/options/server_run_options.go]: done

File: pkg/genericapiserver/serve.go (hunk 12 above)
Comment (sttts) [inline pkg/genericapiserver/serve.go]: Actually, `Server.Serve(listener)` does the http2 setup. The setup call in `ListenAndServeTLS` is superfluous IMO.

File: pkg/genericapiserver/genericapiserver.go (hunk 11 above)
Comment (sttts) [inline pkg/genericapiserver/genericapiserver.go]: good spot.

File: test/integration/discoverysummarizer/discoverysummarizer_test.go (hunk 15 above)
Comment (sttts) [inline test/integration/discoverysummarizer/discoverysummarizer_test.go]: why? It's not used.

Comment (sttts) [inline test/integration/discoverysummarizer/discoverysummarizer_test.go]: And asks for port conflicts. We can set it to 6444 of course.

File: pkg/util/cert/cert.go (hunk 17 above)
Comment (sttts) [inline pkg/util/cert/cert.go]: Would a different serial help then? I don't like to set the common name is something invalid as it is used for SNI.

**Comment by sttts** [comment]
@ncdc @liggitt addressed all comments, ptal

**Comment by sttts** [comment]
@k8s-bot cvm gke e2e test this

File: pkg/util/cert/cert.go (hunk 17 above)
Comment (liggitt) [inline pkg/util/cert/cert.go]: CN is not required to be a hostname in serving certs, and if this is for the default cert, it'll be used for all requests that don't match an explicit cert. We could pick a random serial number instead, but I'd probably leave it as-is. As long as there is a SAN for all the expected domains, the CN can be whatever we want

Comment (sttts) [inline pkg/util/cert/cert.go]: Didn't know CN can be nearly anything (https://tools.ietf.org/html/rfc6125). Will restore the old behaviour.

Comment (sttts) [inline pkg/util/cert/cert.go]: Restored the old behaviour.

**Comment by liggitt** [comment]
LGTM, squash in "address comments" commit

**Review by liggitt — APPROVED** [review]

File: examples/apiserver/apiserver.go [hunk 20]
```diff
@@ -105,6 +106,6 @@ func Run(serverOptions *genericoptions.ServerRunOptions) error {
 	if err := s.InstallAPIGroup(&apiGroupInfo); err != nil {
 		return fmt.Errorf("Error in installing API: %v", err)
 	}
-	s.PrepareRun().Run()
+	s.PrepareRun().Run(wait.NeverStop)
```
Comment (deads2k) [inline examples/apiserver/apiserver.go]: This one we want to stop.

File: pkg/util/cert/cert.go [hunk 21]
```diff
@@ -61,7 +61,10 @@ func NewPrivateKey() (*rsa.PrivateKey, error) {
 func NewSelfSignedCACert(cfg Config, key *rsa.PrivateKey) (*x509.Certificate, error) {
 	now := time.Now()
 	tmpl := x509.Certificate{
-		SerialNumber: new(big.Int).SetInt64(0),
+		// Many clients fail TLS connections if they see two different certs with the same serial number
+		// from the same signer (and they have a local memory of previously seen certs from a given signer).
+		// Hence, we put a timestamp here.
+		SerialNumber: new(big.Int).SetInt64(time.Now().Unix()),
```
Comment (deads2k) [inline pkg/util/cert/cert.go]: I'd rather see UID.  @liggitt did you prefer timestamp?

Comment (deads2k) [inline pkg/util/cert/cert.go]: Oh, and if it switches to uid, it needs to be positive.

**Comment by deads2k** [comment]
minor comments, lgtm otherwise.

File: examples/apiserver/apiserver.go (hunk 20 above)
Comment (sttts) [inline examples/apiserver/apiserver.go]: added a close channel everywhere where the example apiserver is used.

File: pkg/genericapiserver/serve.go [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+		<-stopCh
+		lock.Lock()
+		defer lock.Unlock()
+		ln.Close()
+	}()
+
+	go func() {
+		defer utilruntime.HandleCrash()
+
+		for {
```
Comment (sttts) [inline pkg/genericapiserver/serve.go]: Only after the first synchronous try which worked. That was my understanding from our small irc discussion. A follow-up is fine. I prefer a hard crash in such a case such that the outer orchestrator (e.g. systemd or kubelet) can do its job.

File: pkg/util/cert/cert.go (hunk 21 above)
Comment (sttts) [inline pkg/util/cert/cert.go]: In fact this was an accident that this function was changed at all. It's unrelated to SNI. Will undo the change.

**Comment by sttts** [comment]
Squashed.

File: pkg/genericapiserver/serve.go (hunk 22 above)
Comment (liggitt) [inline pkg/genericapiserver/serve.go]: I can't think of a good reason to retry such a fundamental failure internally
